x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -140,165 +144,122 @@ public String sessionId() {
-                        null
-                    );
-                } else {
-                    exceptionForResponse = e;
-                }
-                for (String clusterAlias : executionInfo.clusterAliases()) {
-                    executionInfo.swapCluster(clusterAlias, (k, v) -> {
-                        EsqlExecutionInfo.Cluster.Builder builder = new EsqlExecutionInfo.Cluster.Builder(v).setTook(
-                            executionInfo.overallTook()
-                        ).setTotalShards(0).setSuccessfulShards(0).setSkippedShards(0).setFailedShards(0);
-                        if (RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY.equals(clusterAlias)) {
-                            // never mark local cluster as skipped
-                            builder.setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL);
-                        } else {
-                            builder.setStatus(EsqlExecutionInfo.Cluster.Status.SKIPPED);
-                            // add this exception to the failures list only if there is no failure already recorded there
-                            if (v.getFailures() == null || v.getFailures().size() == 0) {
-                                builder.setFailures(List.of(new ShardSearchFailure(exceptionForResponse)));
-                            }
-                        }
-                        return builder.build();
-                    });
+            new CcsUtils.CssPartialErrorsActionListener(executionInfo, listener) {
+                @Override
+                public void onResponse(LogicalPlan analyzedPlan) {
+                    executeOptimizedPlan(request, executionInfo, planRunner, optimizedPlan(analyzedPlan), listener);
                 }
-                listener.onResponse(new Result(Analyzer.NO_FIELDS, Collections.emptyList(), Collections.emptyList(), executionInfo));
-            } else {
-                listener.onFailure(e);
             }
-        }
+        );
     }
 
     /**
      * Execute an analyzed plan. Most code should prefer calling {@link #execute} but
-     * this is public for testing. See {@link Phased} for the sequence of operations.
+     * this is public for testing.
      */
     public void executeOptimizedPlan(
         EsqlQueryRequest request,
         EsqlExecutionInfo executionInfo,
-        BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
+        PlanRunner planRunner,
         LogicalPlan optimizedPlan,
         ActionListener<Result> listener
     ) {
-        LogicalPlan firstPhase = Phased.extractFirstPhase(optimizedPlan);
-        updateExecutionInfoAtEndOfPlanning(executionInfo);
-        if (firstPhase == null) {
-            runPhase.accept(logicalPlanToPhysicalPlan(optimizedPlan, request), listener);
+        PhysicalPlan physicalPlan = logicalPlanToPhysicalPlan(optimizedPlan, request);
+        // TODO: this could be snuck into the underlying listener
+        CcsUtils.updateExecutionInfoAtEndOfPlanning(executionInfo);
```

Comment:
@quux00 let me know if this makes sense.

Reply:
The thing to test is to make sure it doesn't throw assertion errors when running with INLINESTATS, since that can do planning twice and this only wants to be called once (until we make it work with the 2 phase execution model). If it passes that, then it's fine for now. Based on reading the code below I think we're OK since the recursion is in `executeSubPlans`.